### PR TITLE
feat: ability tree scroll to navigate

### PR DESCRIPTION
**Pull request by byBackfish**
Ports over [feat: ability tree navigation using scroll wheel #582](https://github.com/Wynntils/Wynntils/pull/582) from legacy to artemis.

Adds ability to use the scroll wheel to navigate through the ability tree

### CODE REVIEW

File: common/src/main/java/com/wynntils/core/features/FeatureRegistry.java [hunk 1]
```diff
@@ -125,6 +126,7 @@ public static void init() {
         registerFeature(new UnidentifiedItemIconFeature());
         registerFeature(new WynncraftButtonFeature());
         registerFeature(new WynncraftPauseScreenFeature());
+        registerFeature(new AbilityTreeScrollFeature());
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/core/features/FeatureRegistry.java]: Maintain alphabetical order please

File: common/src/main/java/com/wynntils/features/user/AddCommandExpansionFeature.java [hunk 2]
```diff
@@ -62,6 +62,7 @@ private void addArgumentlessCommandNodes(RootCommandNode root) {
         root.addChild(literal("hub").build());
         root.addChild(literal("itemlock").build());
         root.addChild(literal("kill").build());
+        root.addChild(literal("hub").build());
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/AddCommandExpansionFeature.java]: Nice catch, but don't mix fixes in PRs next time, please.

File: common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    private final Pattern ABILITY_TREE_PATTERN = Pattern.compile("(?:Warrior|Shaman|Mage|Assassin|Archer) Abilities");
+
+    private static final int abilityTreePreviousSlot = 57;
+    private static final int abilityTreeNextSlot = 59;
+
+    @SubscribeEvent
+    public void onInteract(MouseScrollEvent event) {
+        Screen screen = McUtils.mc().screen;
+        if (screen == null) return;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java]: Why not .matches?

File: common/src/main/java/com/wynntils/mc/mixin/MouseHandlerMixin.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(MouseHandler.class)
+public abstract class MouseHandlerMixin {
+
+    @Inject(method = "onScroll", at = @At("HEAD"))
+    private void onScroll(long windowPointer, double xOffset, double yOffset, CallbackInfo ci) {
+        WynntilsMod.getEventBus().post(new MouseScrollEvent(yOffset > 0));
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/mixin/MouseHandlerMixin.java]: Make a method for this in EventFactory

File: common/src/main/java/com/wynntils/mc/EventFactory.java [hunk 5]
```diff
@@ -411,5 +412,9 @@ public static Event onSubtitleSetText(ClientboundSetSubtitleTextPacket packet) {
         return post(new SubtitleSetTextEvent(packet.getText()));
     }
 
+    public static Event onMouseScroll(double yOffset) {
+        return post(new MouseScrollEvent(yOffset > 0));
```
Comment (magicus) [inline common/src/main/java/com/wynntils/mc/EventFactory.java]: Maybe send in the yOffset as is into the event, and then add a `boolean isScrollingUp() { return yOffset > 0; }` method in the event instead? That makes it a bit more general if anyone wants to use it in the future.

File: common/src/main/java/com/wynntils/features/user/AddCommandExpansionFeature.java (hunk 2 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/features/user/AddCommandExpansionFeature.java]: Also, was it a catch? `hub` was already added, in it's alphabetic place.

File: common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.regex.Pattern;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+@FeatureInfo(stability = FeatureInfo.Stability.STABLE, category = FeatureCategory.INVENTORY)
+public class AbilityTreeScrollFeature extends UserFeature {
+
+    @Config
+    private boolean invertScroll = false;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java]:
```suggestion
    @Config
    public boolean invertScroll = false;
```

Also, move this to be after the private fields.

File: common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    private final Pattern ABILITY_TREE_PATTERN = Pattern.compile("(?:Warrior|Shaman|Mage|Assassin|Archer) Abilities");
+
+    private static final int abilityTreePreviousSlot = 57;
+    private static final int abilityTreeNextSlot = 59;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java]:
```suggestion
    private static final Pattern ABILITY_TREE_PATTERN = Pattern.compile("(?:Warrior|Shaman|Mage|Assassin|Archer) Abilities");
    private static final int abilityTreePreviousSlot = 57;
    private static final int abilityTreeNextSlot = 59;
```

File: common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Config
+    public boolean invertScroll = false;
+
+    @SubscribeEvent
+    public void onInteract(MouseScrollEvent event) {
+        Screen screen = McUtils.mc().screen;
+        if (screen == null) return;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/inventory/AbilityTreeScrollFeature.java]: Do we need this null check if we do the instanceof below?

**Review by kristofbolyai — APPROVED** [review]

**Review by magicus — APPROVED** [review]
Looks good now, methinks. Thanks for implementing this, and porting it to Artemis!

**Comment by kristofbolyai** [comment]
Thanks!

**Pull request merged**